File: Braintree/src/main/java/com/braintreepayments/api/internal/UUIDHelper.java
package com.braintreepayments.api.internal;

import android.content.Context;
import android.content.SharedPreferences;

import java.util.UUID;

/**
 * Contains methods to collect information about the app and device.
 */
public class UUIDHelper {

    private static final String BRAINTREE_UUID_KEY = "braintreeUUID";

    /**
     * @param context
     * @return A persistent UUID for this application install.
     */
    public static String getPersistentUUID(Context context) {
        SharedPreferences prefs = BraintreeSharedPreferences.getSharedPreferences(context);

        String uuid = prefs.getString(BRAINTREE_UUID_KEY, null);
        if (uuid == null) {
            uuid = getFormattedUUID();
            prefs.edit().putString(BRAINTREE_UUID_KEY, uuid).apply();
        }

        return uuid;
    }

    public static String getFormattedUUID() {
        return UUID.randomUUID().toString().replace("-", "");
    }
}

File: BraintreeCore/src/main/java/com/braintreepayments/api/core/UUIDHelper.kt
package com.braintreepayments.api.core

import android.content.Context
import androidx.annotation.RestrictTo
import com.braintreepayments.api.sharedutils.BraintreeSharedPreferences
import java.util.*

/**
 * @suppress
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class UUIDHelper {

    val formattedUUID: String
        get() = UUID.randomUUID().toString().replace("-", "")

    fun getInstallationGUID(context: Context?): String {
        return getInstallationGUID(BraintreeSharedPreferences.getInstance(context))
    }

    internal fun getInstallationGUID(braintreeSharedPreferences: BraintreeSharedPreferences): String {
        var installationGUID = braintreeSharedPreferences.getString(INSTALL_GUID, null)
        if (installationGUID == null) {
            installationGUID = UUID.randomUUID().toString()
            braintreeSharedPreferences.putString(INSTALL_GUID, installationGUID)
        }
        return installationGUID
    }

    companion object {
        private const val INSTALL_GUID = "InstallationGUID"
    }
}
